code/controllers/subsystem/blackbox.dm

```diff
@@ -0,0 +1,274 @@
+		if(FV.key == key)
+			return FV
+
+	var/datum/feedback_variable/FV = new(key, key_type)
+	feedback += FV
+	return FV
+
+/*
+feedback data can be recorded in 5 formats:
+"text"
+	used for simple single-string records i.e. the current map
+	further calls to the same key will append saved data unless the overwrite argument is true or it already exists
+	when encoded calls made with overwrite will lack square brackets
+	calls: 	SSblackbox.record_feedback("text", "example", 1, "sample text")
+			SSblackbox.record_feedback("text", "example", 1, "other text")
+	json: {"data":["sample text","other text"]}
+"amount"
+	used to record simple counts of data i.e. the number of ahelps recieved
+	further calls to the same key will add or subtract (if increment argument is a negative) from the saved amount
+	calls:	SSblackbox.record_feedback("amount", "example", 8)
+			SSblackbox.record_feedback("amount", "example", 2)
+	json: {"data":10}
+"tally"
+	used to track the number of occurances of multiple related values i.e. how many times each type of gun is fired
+	further calls to the same key will:
+	 	add or subtract from the saved value of the data key if it already exists
+		append the key and it's value if it doesn't exist
+	calls:	SSblackbox.record_feedback("tally", "example", 1, "sample data")
+			SSblackbox.record_feedback("tally", "example", 4, "sample data")
+			SSblackbox.record_feedback("tally", "example", 2, "other data")
+	json: {"data":{"sample data":5,"other data":2}}
+"nested tally"
+	used to track the number of occurances of structured semi-relational values i.e. the results of arcade machines
+	similar to running total, but related values are nested in a multi-dimensional array built
+	the final element in the data list is used as the tracking key, all prior elements are used for nesting
+	further calls to the same key will:
+	 	add or subtract from the saved value of the data key if it already exists in the same multi-dimensional position
+		append the key and it's value if it doesn't exist
+	calls: 	SSblackbox.record_feedback("nested tally", "example", 1, list("fruit", "orange", "apricot"))
+			SSblackbox.record_feedback("nested tally", "example", 2, list("fruit", "orange", "orange"))
+			SSblackbox.record_feedback("nested tally", "example", 3, list("fruit", "orange", "apricot"))
+			SSblackbox.record_feedback("nested tally", "example", 10, list("fruit", "red", "apple"))
+			SSblackbox.record_feedback("nested tally", "example", 1, list("vegetable", "orange", "carrot"))
+	json: {"data":{"fruit":{"orange":{"apricot":4,"orange":2},"red":{"apple":10}},"vegetable":{"orange":{"carrot":1}}}}
+	tracking values associated with a number can't merge with a nesting value, trying to do so will append the list
+	call:	SSblackbox.record_feedback("nested tally", "example", 3, list("fruit", "orange"))
+	json: {"data":{"fruit":{"orange":{"apricot":4,"orange":2},"red":{"apple":10},"orange":3},"vegetable":{"orange":{"carrot":1}}}}
```

Comment:
This is very bad, because *depending on the tool you use to parse the json* you'll be getting different results.
Can we have it runtime instead/in addition to this?

Reply:
Could you elaborate?

Reply:
Some tools/languages/libraries would error out trying to parse `{"data":{"fruit":{"orange":{"apricot":4,"orange":2},"red":{"apple":10},"orange":3}`;
some would treat it  as `{"data":{"fruit":{"orange":{"apricot":4,"orange":2},"red":{"apple":10}}`;
some as `{"data":{"fruit":{"red":{"apple":10},"orange":3}`
some will treat it as either of the above randomly in each individual run.
This nondeterminism is bad and makes for hard to catch bugs..

Reply:
Ooh, I see the problem. The concept of nested tallies itself isn't a problem, it's the fact that this leads to duplicated keys. Note the warning given.
![https://i.tigercat2000.net/2020/12/firefox_LBrP5WTtJN.png](https://i.tigercat2000.net/2020/12/firefox_LBrP5WTtJN.png)

Reply:
Wait, is this even possible in BYOND? Associative lists have unique keys, how would this kind of JSON be represented internally?